docs: added prompt webhook tutorial (#806)
langchain-ai/langsmith-docs: docs/prompt_engineering/tutorials/prompt_commit.mdx

A powerful feature to enhance this workflow is the ability to react to prompt modifications using **webhooks**. When a prompt is updated in Langsmith and a "prompt commit" is made, a webhook can trigger external actions. This tutorial will guide you through setting up a system to automatically commit your Langsmith prompt manifest to a GitHub repository whenever a prompt is updated.

**Why sync prompts with GitHub?**

* **Version Control:** Keep your prompts versioned alongside your application code in a familiar system.
* **CI/CD Integration:** Trigger automated staging or production deployments when critical prompts change.
* **Auditing & Collaboration:** Maintain a clear audit trail of who changed what and when, facilitating team collaboration even with non-technical members who might be editing prompts in the Langsmith UI.
* **Codebase Synchronization:** Ensure your application's codebase always has access to the latest approved prompt versions.
* **Automated Testing & Evaluation:** Automatically re-run evaluation datasets or integration tests when a prompt is modified to catch regressions or validate improvements.

> * **Documentation:** Automatically update internal or external documentation related to prompt structures or purposes.

Reviewer: I would get rid of these ones

## Prerequisites

Before we begin, ensure you have the following set up:

1.  **Langsmith Account:** You'll need access to a [Langsmith](https://smith.langchain.com/) workspace.
2.  **GitHub Account:** A standard GitHub account.
3.  **GitHub Repository:** Create a new (or choose an existing) repository where your Langsmith prompt manifests will be stored. This could be the same repository as your application code or a dedicated one for prompts.
4.  **GitHub Personal Access Token (PAT):**

>     * Langsmith webhooks themselves don't directly interact with GitHub; instead, they'll call an intermediary server that *you* create. This server will need a GitHub PAT to authenticate and make commits to your repository.

Reviewer 2: What do you think of adding a sequence diagram after the `The Workflow in Action` title as well?

```mermaid
sequenceDiagram
    participant User
    participant LangSmith
    participant WebhookServer as FastAPI Server
    participant GitHub

    User->>LangSmith: Save prompt (creates prompt commit)
    LangSmith-->>WebhookServer: POST webhook with manifest JSON
    WebhookServer->>GitHub: GET file (to fetch current SHA)
    GitHub-->>WebhookServer: 200 OK (or 404 if file not found)
    WebhookServer->>GitHub: PUT file (commit manifest update)
    GitHub-->>WebhookServer: 201 Created (commit details)
    WebhookServer-->>LangSmith: 201 Created (acknowledge webhook)
```

>     * Langsmith webhooks themselves don't directly interact with GitHub; instead, they'll call an intermediary server that *you* create. This server will need a GitHub PAT to authenticate and make commits to your repository.

Reviewer: I like this!!

>     * Langsmith webhooks themselves don't directly interact with GitHub; instead, they'll call an intermediary server that *you* create. This server will need a GitHub PAT to authenticate and make commits to your repository.

Reviewer 2: When adding the mermaid code with triple backticks I get this error in the UI, I wouldn't want to change docusaurus configuration tho
```python
Uncaught runtime errors:
×
ERROR
Hook useColorMode is called outside the <ColorModeProvider>. Please see https://docusaurus.io/docs/api/themes/configuration#use-color-mode.
ReactContextError
    at useColorMode (webpack-internal:///./node_modules/@docusaurus/theme-mermaid/node_modules/@docusaurus/theme-common/lib/contexts/colorMode.js:27:1533)
    at useMermaidConfig (webpack-internal:///./node_modules/@docusaurus/theme-mermaid/lib/client/index.js:19:303)
    at useMermaidRenderResult (webpack-internal:///./node_modules/@docusaurus/theme-mermaid/lib/client/index.js:45:35)
    at MermaidRenderer (webpack-internal:///./node_modules/@docusaurus/theme-mermaid/lib/theme/Mermaid/index.js:18:196)
    at renderWithHooks (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:15486:18)
    at mountIndeterminateComponent (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:20098:13)
    at beginWork (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:21621:16)
    at HTMLUnknownElement.callCallback (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:4213:16)
    at invokeGuardedCallback (webpack-internal:///./node_modules/react-dom/cjs/react-dom.development.js:4277:31)
```

>     * Langsmith webhooks themselves don't directly interact with GitHub; instead, they'll call an intermediary server that *you* create. This server will need a GitHub PAT to authenticate and make commits to your repository.

Reviewer 2: Added an image for now

    * **Permissions:** The PAT needs the `repo` scope (or at least `public_repo` if your repository is public, and `repo:status`, `repo_deployment`, `public_repo` for full access to commit content).
    * **Generation:**
        * Go to your GitHub **Settings**.
        * Navigate to **Developer settings** &gt; **Personal access tokens** &gt; **Tokens (classic)**.
        * Click **"Generate new token"** (or "Generate new token (classic)").
        * Give it a descriptive name (e.g., "Langsmith Prompt Sync").
        * Set an expiration period.
        * Select the `repo` scope.
        * Click **"Generate token"**.
        * **Important:** Copy the token immediately. You won't be able to see it again. Store it securely, as we'll use it as an environment variable for our server.

> 5.  **Publicly Accessible Webhook Receiver (FastAPI Server):**

Reviewer 2: Yes I agree! I will move the FastAPI server part in a new section before `Configuring the Webhook in LangSmith` one.

```python
import base64
import json
from typing import Any, Dict

import httpx
from fastapi import FastAPI, HTTPException, Body
from pydantic import BaseModel, Field
from pydantic_settings import BaseSettings, SettingsConfigDict

# --- Configuration ---
class AppConfig(BaseSettings):
    """
    Application configuration model.
    Loads settings from environment variables.
    """
    GITHUB_TOKEN: str
    GITHUB_REPO_OWNER: str
    GITHUB_REPO_NAME: str
    GITHUB_FILE_PATH: str = "prompt_manifest.json"
    GITHUB_BRANCH: str = "main"

    model_config = SettingsConfigDict(
        env_file=".env",
        env_file_encoding='utf-8',
        extra='ignore'
    )

settings = AppConfig()

# --- Pydantic Models ---
class WebhookPayload(BaseModel):
    """
    Defines the expected structure of the incoming webhook payload.
    """
    manifest: Dict[str, Any] = Field(
        ...,
        description="The main content or configuration data to be committed to GitHub."
    )
    commit_hash: str = Field(
        ...,
        description="An identifier for the commit event that triggered the webhook."
    )
    created_at: str = Field(
        ...,

>         description="Timestamp indicating when the event was created (ISO format preferred)."

Reviewer: add prompt_name, created_by, prompt_id

# 2. Create a .env file with your GitHub token and repo details.
# 3. Run with Uvicorn: uvicorn main:app --reload
# 4. Deploy to a public platform like Render.com.
```

**Key aspects of this server:**
* **Configuration (`.env`):** It expects a `.env` file with your `GITHUB_TOKEN`, `GITHUB_REPO_OWNER`, and `GITHUB_REPO_NAME`. You can also customize `GITHUB_FILE_PATH` (default: `langsmith_prompt_manifest.json`) and `GITHUB_BRANCH` (default: `main`).
* **GitHub Interaction:** The `commit_manifest_to_github` function handles the logic of fetching the current file's SHA (to update it) and then committing the new manifest content.
* **Webhook Endpoint (`/webhook/commit`):** This is the URL path your Langsmith webhook will target.
* **Error Handling:** Basic error handling for GitHub API interactions is included.

**Deploy this server to your chosen platform (e.g., Render) and note down its public URL (e.g., `https://prompt-commit-webhook.onrender.com`).**

</details>

## Understanding Langsmith "Prompt Commits" and Webhooks

In Langsmith, when you save changes to a prompt, you're essentially creating a new version or a "Prompt Commit." These commits are what can trigger webhooks.

> The webhook will typically send a JSON payload containing the entire **prompt manifest** for your workspace. This manifest includes all the prompts defined in that workspace.

Reviewer: ```suggestion
The webhook will send a JSON payload containing the new **prompt manifest**.
```

<details>
<summary>Sample Webhook Payload</summary>

```json
{
  "prompt_id": "f33dcb51-eb17-47a5-83ca-64ac8a027a29",
  "commit_hash": "commit_hash_1234567890",
  "created_at": "2021-01-01T00:00:00Z",
  "created_by": "Jane Doe",

>   "manifest": {

Reviewer: add prompt_name

                    "template": "{{question}}"
                  }
                }
              }
            }
          ],
          "input_variables": [
            "question"
          ]
        }
      },
      "last": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "runnable",
          "RunnableBinding"
        ],
        "kwargs": {
          "bound": {
            "lc": 1,
            "type": "constructor",
            "id": [
              "langchain",
              "chat_models",
              "openai",
              "ChatOpenAI"
            ],
            "kwargs": {
              "temperature": 1,
              "top_p": 1,
              "presence_penalty": 0,
              "frequency_penalty": 0,
              "model": "gpt-4.1-mini",
              "extra_headers": {},
              "openai_api_key": {
                "id": [
                  "OPENAI_API_KEY"
                ],
                "lc": 1,
                "type": "secret"
              }
            }
          },
          "kwargs": {}
        }
      }
    }
  }
}
```

</details>

:::note Workspace Level Triggers

> It's important to understand that Langsmith webhooks for prompt commits are generally triggered at the **workspace level**. This means if *any* prompt within your Langsmith workspace is modified and a "prompt commit" is saved, the webhook will fire and send the *entire manifest* of all prompts in that workspace. Your receiving server should be designed with this in mind.

Reviewer: ```suggestion
It's important to understand that Langsmith webhooks for prompt commits are generally triggered at the **workspace level**. This means if *any* prompt within your Langsmith workspace is modified and a "prompt commit" is saved, the webhook will fire and send the updated manifest of the prompt. The payloads are identifiable by prompt id. Your receiving server should be designed with this in mind.
```

2.  Go to the **Prompts** section. Here you'll see a list of your prompts.

    ![LangSmith Prompts section](./static/prompt-commit-main.png)
    *(Illustrative image: Langsmith Prompts section)*

3.  On the top right of the Prompts page, click the **+ Webhook** button.

4.  You'll be presented with a form to configure your webhook:

    ![Langsmith Webhook configuration modal](./static/prompt-commit-webhook.png)
    *(Illustrative image: Langsmith Webhook configuration modal)*

    * **Webhook URL:** Enter the full public URL of your deployed FastAPI server's endpoint. For our example server, this would be `https://your-app-name.onrender.com/webhook/commit`.
    * **Headers (Optional):**
        * You can add custom headers that Langsmith will send with each webhook request.

5.  **Test the Webhook:** Langsmith provides a "Send Test Notification" button. Use this to send a sample payload to your server. Check your server logs (e.g., on Render) to ensure it receives the request and processes it successfully (or to debug any issues).

6.  **Save** the webhook configuration.

## The Workflow in Action

Now, with everything set up, here's what happens:

> 1.  **Prompt Modification:** A user (developer or non-technical team member) modifies a prompt in the Langsmith UI and saves it, creating a new "prompt commit."

Reviewer 3: ```suggestion
1.  **Prompt Modification:** A user modifies a prompt in the Langsmith UI and saves creating a new "prompt commit."
```

2.  **Webhook Trigger:** Langsmith detects this new prompt commit and triggers the configured webhook.
3.  **HTTP Request:** Langsmith sends an HTTP POST request to the public URL of your FastAPI server (e.g., `https://prompt-commit-webhook.onrender.com/webhook/commit`). The body of this request contains the JSON prompt manifest for the entire workspace.
4.  **Server Receives Payload:** Your FastAPI server's endpoint receives the request.
5.  **GitHub Commit:** The server parses the JSON manifest from the request body. It then uses the configured GitHub Personal Access Token, repository owner, repository name, file path, and branch to:
    * Check if the manifest file already exists in the repository on the specified branch to get its SHA (this is necessary for updating an existing file).
    * Create a new commit with the latest prompt manifest, either creating the file or updating it if it already exists. The commit message will indicate that it's an update from Langsmith.

> 6.  **Confirmation:** Your server sends a success response back to Langsmith. You should see the new commit appear in your GitHub repository.

Reviewer 3: "Your server sends a success response back to Langsmith"

is this true?

Reviewer 2: Yes, generally I’d expect the customer’s server to return a 2xx status code to acknowledge the webhook (that’s typical for RESTful endpoints). But let’s check with @madams0013 in case we handle the response differently on our end

Reviewer: I didn't do anything to handle the success response, I just log any errors -- maybe we should add this though?

Reviewer 2: > I didn't do anything to handle the success response, I just log any errors -- maybe we should add this though? in the ui I just say `success` if the webhook is sent successfully

Yeah, I tried adding a random url and it still sends the webhook and notifies that it has been successful. Maybe we should wait for the user response and eventually throw an error on timeout (like 10 sec)?

Reviewer 2: removed that sentence for now!
